cmd/kubeadm/app/master/discovery.go

```diff
@@ -0,0 +1,136 @@
+
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/api"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	certutil "k8s.io/kubernetes/pkg/util/cert"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+type kubeDiscovery struct {
+	Deployment *extensions.Deployment
+	Secret     *api.Secret
+}
+
+const (
+	kubeDiscoveryName       = "kube-discovery"
+	kubeDiscoverySecretName = "clusterinfo"
+)
+
+func encodeKubeDiscoverySecretData(s *kubeadmapi.KubeadmConfig, caCert *x509.Certificate) map[string][]byte {
+	var (
+		data         = map[string][]byte{}
+		endpointList = []string{}
+		tokenMap     = map[string]string{}
+	)
+
+	for _, addr := range s.InitFlags.API.AdvertiseAddrs {
+		endpointList = append(endpointList, fmt.Sprintf("https://%s:443", addr.String()))
+	}
+
+	tokenMap[s.Secrets.TokenID] = s.Secrets.BearerToken
+
+	data["endpoint-list.json"], _ = json.Marshal(endpointList)
+	data["token-map.json"], _ = json.Marshal(tokenMap)
+	data["ca.pem"] = certutil.EncodeCertPEM(caCert)
+
+	return data
+}
+
+func newKubeDiscoveryPodSpec(s *kubeadmapi.KubeadmConfig) api.PodSpec {
+	return api.PodSpec{
+		// We have to use host network namespace, as `HostPort`/`HostIP` are Docker's
+		// buisness and CNI support isn't quite there yet (except for kubenet)
+		// (see https://github.com/kubernetes/kubernetes/issues/31307)
+		// TODO update this when #31307 is resolved
+		SecurityContext: &api.PodSecurityContext{HostNetwork: true},
+		Containers: []api.Container{{
+			Name:    kubeDiscoveryName,
+			Image:   s.EnvParams["discovery_image"],
+			Command: []string{"/usr/local/bin/kube-discovery"},
+			VolumeMounts: []api.VolumeMount{{
+				Name:      kubeDiscoverySecretName,
+				MountPath: "/tmp/secret", // TODO use a shared constant
+				ReadOnly:  true,
+			}},
+			Ports: []api.ContainerPort{
+				// TODO when CNI issue (#31307) is resolved, we should consider adding
+				// `HostIP: s.API.AdvertiseAddrs[0]`, if there is only one address`
+				{Name: "http", ContainerPort: 9898, HostPort: 9898},
+			},
+		}},
+		Volumes: []api.Volume{{
+			Name: kubeDiscoverySecretName,
+			VolumeSource: api.VolumeSource{
+				Secret: &api.SecretVolumeSource{SecretName: kubeDiscoverySecretName},
+			}},
+		},
+	}
+}
+
+func newKubeDiscovery(s *kubeadmapi.KubeadmConfig, caCert *x509.Certificate) kubeDiscovery {
+	kd := kubeDiscovery{
+		Deployment: NewDeployment(kubeDiscoveryName, 1, newKubeDiscoveryPodSpec(s)),
+		Secret: &api.Secret{
+			ObjectMeta: api.ObjectMeta{Name: kubeDiscoverySecretName},
+			Type:       api.SecretTypeOpaque,
+			Data:       encodeKubeDiscoverySecretData(s, caCert),
+		},
+	}
+
+	SetMasterTaintTolerations(&kd.Deployment.Spec.Template.ObjectMeta)
+	SetMasterNodeAffinity(&kd.Deployment.Spec.Template.ObjectMeta)
+
+	return kd
+}
+
+func CreateDiscoveryDeploymentAndSecret(s *kubeadmapi.KubeadmConfig, client *clientset.Clientset, caCert *x509.Certificate) error {
```

Comment:
> Why isn't this a static pod?

It uses a secret, which depends on control plane, it doesn't make any
sense to split these two things. However, the main principle driving this
decision was to mod away from static pods where possible. Not using secret is also bad, as we have to write more files to disk, and I believe it's something we prefer to avoid (especially for new pieces). We can change how `kube-discovery` is configured, e.g. be use flags, but that'd make token rotation problem look different too.

> It seems like it would make the transition into core apiserver functionality more straightforward.

The status of whether it may become part the apiserver is still uncertain.